### ISSUE TITLE
[@typescript-eslint/keyword-spacing] Expected space(s) before "this"

### ISSUE THREAD

**Issue by MaxGraey** [typescript-eslint/typescript-eslint#2630]
Got false positive error like `error  Expected space(s) before "this"  @typescript-eslint/keyword-spacing` which triggered on such examples:
```ts
return <Boo>this.foo();
```
I expected `this` in this context will be just ignored.
I know better use "this.foo() as Boo" but this not possible unfortunately. 

I guess similar to #1739 better exclude "this" similar to "as" keyword for such cases or resolve it in more general level.

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/keyword-spacing": "error"
  }
}
```
I trigger with `"@typescript-eslint/keyword-spacing": ["error", { "before": flase }]` as well.
And unfortunately excluding via overrides "this" doesn't work:
```JSON
{
  "rules": {
    "@typescript-eslint/keyword-spacing": ["error", { "overrides": {
       "this": { "before": false }
    }}]
  }
}
```

```TS
// your repro code case
```

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.3.0` |
| `@typescript-eslint/parser`        | `4.3.0` |
| `TypeScript`                       | `4.0.2` |
| `ESLint`                           | `7.10.0` |
| `node`                             | `14.13.0` |

**Comment by bradzacher** [comment]
out of curiosity, why does your project use angle-bracket assertions instead of `as` assertions?
IMO angle bracket assertions are a worse syntax as they look similar to generics and JSX.

**Comment by MaxGraey** [comment]
> IMO angle bracket assertions are a worse syntax as they look similar to generics and JSX

Yes, I know. And mentioned about this. The main reason is speed of parsing. We use our own parser and tokenizer which more prefer `<B>a` than `a as B` in term of parsing speed. Anyway I guess `<B>a` should properly work for eslint as well

**Comment by MaxGraey** [comment]
It seems this is parser problem. See [eslint.org/demo fiddle](https://eslint.org/demo#eyJ0ZXh0IjoiZXhwb3J0IGZ1bmN0aW9uIG9rKGEpIHtcbiAgcmV0dXJuIGE8dGhpcy5hO1xufVxuXG5leHBvcnQgZnVuY3Rpb24gZmFpbChhKSB7XG4gIHJldHVybiBhPnRoaXMuYTtcbn0iLCJvcHRpb25zIjp7InBhcnNlck9wdGlvbnMiOnsiZWNtYVZlcnNpb24iOjExLCJzb3VyY2VUeXBlIjoibW9kdWxlIiwiZWNtYUZlYXR1cmVzIjp7fX0sInJ1bGVzIjp7ImNvbnN0cnVjdG9yLXN1cGVyIjoyLCJmb3ItZGlyZWN0aW9uIjoyLCJnZXR0ZXItcmV0dXJuIjoyLCJuby1hc3luYy1wcm9taXNlLWV4ZWN1dG9yIjoyLCJuby1jYXNlLWRlY2xhcmF0aW9ucyI6Miwibm8tY2xhc3MtYXNzaWduIjoyLCJuby1jb21wYXJlLW5lZy16ZXJvIjoyLCJuby1jb25kLWFzc2lnbiI6Miwibm8tY29uc3QtYXNzaWduIjoyLCJuby1jb25zdGFudC1jb25kaXRpb24iOjIsIm5vLWNvbnRyb2wtcmVnZXgiOjIsIm5vLWRlYnVnZ2VyIjoyLCJuby1kZWxldGUtdmFyIjoyLCJuby1kdXBlLWFyZ3MiOjIsIm5vLWR1cGUtY2xhc3MtbWVtYmVycyI6Miwibm8tZHVwZS1lbHNlLWlmIjoyLCJuby1kdXBlLWtleXMiOjIsIm5vLWR1cGxpY2F0ZS1jYXNlIjoyLCJuby1lbXB0eSI6Miwibm8tZW1wdHktY2hhcmFjdGVyLWNsYXNzIjoyLCJuby1lbXB0eS1wYXR0ZXJuIjoyLCJuby1leC1hc3NpZ24iOjIsIm5vLWV4dHJhLWJvb2xlYW4tY2FzdCI6Miwibm8tZXh0cmEtc2VtaSI6Miwibm8tZmFsbHRocm91Z2giOjIsIm5vLWZ1bmMtYXNzaWduIjoyLCJuby1nbG9iYWwtYXNzaWduIjoyLCJuby1pbXBvcnQtYXNzaWduIjoyLCJuby1pbm5lci1kZWNsYXJhdGlvbnMiOjIsIm5vLWludmFsaWQtcmVnZXhwIjoyLCJuby1pcnJlZ3VsYXItd2hpdGVzcGFjZSI6Miwibm8tbWlzbGVhZGluZy1jaGFyYWN0ZXItY2xhc3MiOjIsIm5vLW1peGVkLXNwYWNlcy1hbmQtdGFicyI6Miwibm8tbmV3LXN5bWJvbCI6Miwibm8tb2JqLWNhbGxzIjoyLCJuby1vY3RhbCI6Miwibm8tcHJvdG90eXBlLWJ1aWx0aW5zIjoyLCJuby1yZWRlY2xhcmUiOjIsIm5vLXJlZ2V4LXNwYWNlcyI6Miwibm8tc2VsZi1hc3NpZ24iOjIsIm5vLXNldHRlci1yZXR1cm4iOjIsIm5vLXNoYWRvdy1yZXN0cmljdGVkLW5hbWVzIjoyLCJuby1zcGFyc2UtYXJyYXlzIjoyLCJuby10aGlzLWJlZm9yZS1zdXBlciI6Miwibm8tdW5kZWYiOjIsIm5vLXVuZXhwZWN0ZWQtbXVsdGlsaW5lIjoyLCJuby11bnJlYWNoYWJsZSI6Miwibm8tdW5zYWZlLWZpbmFsbHkiOjIsIm5vLXVuc2FmZS1uZWdhdGlvbiI6Miwibm8tdW51c2VkLWxhYmVscyI6Miwibm8tdW51c2VkLXZhcnMiOjIsIm5vLXVzZWxlc3MtY2F0Y2giOjIsIm5vLXVzZWxlc3MtZXNjYXBlIjoyLCJuby13aXRoIjoyLCJyZXF1aXJlLXlpZWxkIjoyLCJ1c2UtaXNuYW4iOjIsInZhbGlkLXR5cGVvZiI6Miwia2V5d29yZC1zcGFjaW5nIjoyfSwiZW52Ijp7fX19)

**Comment by bradzacher** [comment]
> It seems this is parser problem. See eslint.org/demo fiddle

No. the eslint fiddle just doesn't use a ts-compatible parser.

> in term of parsing speed

I'm certainly no expert in parsers (at all), but to me it seems like you're talking about _nanoseconds_ of difference between the two.
IMO code quality and developer efficiency >>>>>> parsing perf.
But hey - you do you 😄 

> Anyway I guess <B>a should properly work for eslint as well

Yup, which is why I marked this as a bug.
Happy to accept a PR.

Most people don't use angle bracket syntax, so your best bet to get this fixed would be diving in and fixing this yourself 😃

**Comment by MaxGraey** [comment]
> No. the eslint fiddle just doesn't use a ts-compatible parser.

Yes, but this happened without `jsx` flag in pure JS eslint. I guess typescript-eslint just borrow this behaviour or it use completely different parser?

**Comment by yeonjuan** [comment]
@MaxGraey HI:)

>Got false positive error like error Expected space(s) before "this" @typescript-eslint/keyword-spacing which triggered on such examples:
return <Boo>this.foo();
> {
  "rules": {
    "@typescript-eslint/keyword-spacing": "error"
  }
}

Error on this configuration seems not a false positive. because the default option "before"  is "true". - [keyword-spacing options](https://eslint.org/docs/rules/keyword-spacing#options)

> I trigger with "@typescript-eslint/keyword-spacing": ["error", { "before": flase }] as well.
And unfortunately excluding via overrides "this" doesn't work:
{
  "rules": {
    "@typescript-eslint/keyword-spacing": ["error", { "overrides": {
       "this": { "before": false }
    }}]
  }
}

It's not occurred on the current master branch ([PR confirming it](https://github.com/typescript-eslint/typescript-eslint/pull/2632/files)). Could you check your eslint config or versions(eslint, typescript-eslint...)?

**Comment by MaxGraey** [comment]
> Error on this configuration seems not a false positive. because the default option "before" is "true"

That's correct. But main idea is ignore only `this` and don't touch other keywords. If you add:

`overrides: { this: { before: false } }`

it will be fail in this situation:
```ts
if (a) this.foo(a);
    // ^ Unexpected space(s) before "this"
```

**Comment by yeonjuan** [comment]
@MaxGraey Right It was my misunderstanding Thanks for the explaining.

**Comment by fy0** [comment]
Is any solution exists now?

**Comment by armano2** [comment]
> I trigger with "@typescript-eslint/keyword-spacing": ["error", { "before": flase }] as well.
> And unfortunately excluding via overrides "this" doesn't work:

i did some test in latest version of plugin and it seem to be working now a expected 

```ts
/*eslint "@typescript-eslint/keyword-spacing": ["error", { "before": false }]*/

if (a) this.foo(a);
//    ^ reports as unexpected
return <Boo> this.foo();
//          ^ reports as unexpected
```
```ts
/*eslint "@typescript-eslint/keyword-spacing": ["error", { "before": true }]*/

if (a) this.foo(a);
return <Boo> this.foo();
// no errors reported
```
```ts
/*eslint "@typescript-eslint/keyword-spacing": ["error", { "overrides": { "this": { "before": false } }}]*/

if (a) this.foo(a);
//    ^ reports as unexpected
return <Boo> this.foo();
//          ^ reports as unexpected
```
```ts
/*eslint "@typescript-eslint/keyword-spacing": ["error", { "overrides": { "this": { "before": true } }}]*/

if (a) this.foo(a);
return <Boo> this.foo();
// no errors reported
```

maybe i'm understunding an issue incorrectly but it seem to be working fine now
